### DISCUSSION TITLE
Classifier.predict() says my pretrained and accurate model is inaccurate.

### DISCUSSION THREAD

**Discussion by dwight-nwaigwe** [Trusted-AI/adversarial-robustness-toolbox#1628] [Q&A] [answered]
**Describe the bug**

I included a code snippet. Basically, I use the example located at <https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/examples/get_started_tensorflow.py>, with some variations. Instead of training a model, I create my model and initialize with weights loaded from a pickle file. To verify my model works properly, I call classifier.predict(). Instead of getting a high percentage, i get 8%-10% accuracy, which is not in agreement with the test accuracy I find when I don't use classifier.predict(). For some reason unknown to me, classifier.predict() gives me numbers indicative of random guesses. I assume that my pretrained model is not correctly being read by the aversarial robustness toolbox framework.

**To Reproduce**

`
import tensorflow.compat.v1 as tf
import numpy as np
tf.compat.v1.disable_eager_execution()
#from tensorflow.compat.v1.keras.models import Model
import tensorflow.compat.v1.keras.layers
from tensorflow.compat.v1.keras.models import Sequential
from tensorflow.compat.v1.keras.layers import Dense, Dropout, Activation, Flatten
from art.estimators.classification import TensorFlowClassifier
from art.utils import load_mnist
import pickle

### Step 1: Load the MNIST dataset  

(x_train, y_train), (x_test, y_test), min_pixel_value, max_pixel_value = load_mnist()

# load layer weights
with open('C:\\Users\\valentin\\Desktop\\my_name\\file_adv.pkl', 'rb') as f:
    lay_weights = pickle.load(f)
    
    
# Step 2: Create the model
input_ph = tf.placeholder(tf.float32, shape=[None, 28, 28, 1])
labels_ph = tf.placeholder(tf.int32, shape=[None, 10])

model = Sequential()
model.add(Flatten(input_shape=(28, 28, 1)))
model.add(Activation('relu'))
model.add(Dense(300))
model.add(Activation('relu'))
model.add(Dense(200))
model.add(Activation('relu'))
model.add(Dense(150))
model.add(Activation('relu'))
model.add(Dense(150))
model.add(Activation('relu'))
model.add(Dense(100))
model.add(Activation('sigmoid'))
model.add(Dense(10))
model.layers[2].set_weights([lay_weights[0], lay_weights[1]])
model.layers[4].set_weights([lay_weights[2], lay_weights[3]])
model.layers[6].set_weights([lay_weights[4], lay_weights[5]])
model.layers[8].set_weights([lay_weights[6], lay_weights[7]])
model.layers[10].set_weights([lay_weights[8], lay_weights[9]])
model.layers[12].set_weights([lay_weights[10], lay_weights[11]])

#set logits equal to model output
logits=model(input_ph)

#check that my model is accurate
for ind in range(0,10):
    x_sample1=x_test[ind,:,:,:]
    x_sample1=np.expand_dims(x_sample1,0)
    mod_output=model.predict(x_sample1)
    print(np.argmax(mod_output))
    print(np.argmax(y_test[ind]))

sess = tf.Session()
sess.run(tf.global_variables_initializer())

# Step 3: Create the ART classifier

classifier = TensorFlowClassifier(
    clip_values=(min_pixel_value, max_pixel_value),
    input_ph=input_ph,
    output=logits,
    labels_ph=labels_ph,
    learning=None,
    sess=sess,
    preprocessing_defences=[],
)

# Step 5: Evaluate the ART classifier on benign test examples
predictions = classifier.predict(x_test)
accuracy = np.sum(np.argmax(predictions, axis=1) == np.argmax(y_test, axis=1)) / len(y_test)
print("Accuracy on benign test examples: {}%".format(accuracy * 100))

`
**Expected behavior**
I expect classifier.predict() to report an accuracy (high 90th percentile) of my model that is in agreement with my other tests.

**Screenshots**

**System information (please complete the following information):**
 - Windows 10
 - python 3.7.10 
 - ART version or commit number
 - TensorFlow 2.1 and 2.4.1

**Comment by beat-buesser** [comment, marked as answer]
Hi @animalcroc Are you able to extend/modify this script to make correct predictions just with your model only?

**Reply by dwight-nwaigwe** [reply]
Hi @beat-buesser, I do not understand your question. The script I attached has a section where I test the model predictions. Is this what you are looking for?

**Reply by beat-buesser** [reply]
@animalcroc Yes, I was referring to the section after the comment `#check that my model is accurate` in Step 2. What is the accuracy on the test data using `mod_output=model.predict(x_sample1)` on all images in `x_test`? I'm asking as first step to see if your model works as expected.

**Reply by dwight-nwaigwe** [reply]
@beat-buesser  It is approximately 97% accurate.

**Reply by beat-buesser** [reply]
@animalcroc Could it be possible that `sess.run(tf.global_variables_initializer())` overwrites the loaded weights? Can you compare the weights of the layers before and after this line and check if they are the same?

**Reply by dwight-nwaigwe** [reply]
@beat-buesser I compared the model accuracy before and after this line, and the accuracy is unchanged.  The problem only occurs when the classifier.predict() function outputs the accuracy.

**Reply by beat-buesser** [reply]
@animalcroc Can you share the layer weights to reproduce the script?

**Reply by dwight-nwaigwe** [reply]
@beat-buesser Sure, but how do I do that? I cannot attach a pickle file nor a  zip file. I don't see the allowable file types.

**Reply by beat-buesser** [reply]
@animalcroc In the field where you type the reply, at the bottom, click on "Attach files by dragging, selecting or pasting them"

**Reply by dwight-nwaigwe** [reply]
@beat-buesser @My file is in pickle (.pkl) format, and I can't upload it, even if i make it .zip file. Any suggestions?

**Reply by beat-buesser** [reply]
@animalcroc What does the pkl-file contain, is it a Numpy array?

**Reply by beat-buesser** [reply]
Can you show how you are saving the model weights?

**Reply by dwight-nwaigwe** [reply]
@beat-buesser  I attached a text file containing the layer weights as a list. Upon successfully reading of this list from the file, you can use my code (attached in the first post) to assign the weights to the layers.

[layer_weights.txt](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/8501151/layer_weights.txt)

**Reply by beat-buesser** [reply]
@animalcroc That could work, but you would have to save the entire array. In the current text file some of the arrays are truncated. Try
```python
import sys
import numpy
numpy.set_printoptions(threshold=sys.maxsize)
```
from https://stackoverflow.com/questions/1987694/how-to-print-the-full-numpy-array-without-truncation

**Reply by beat-buesser** [reply]
@animalcroc Could you also please share the code to calculate 97% accuracy with the loaded weights?

**Reply by dwight-nwaigwe** [reply]
@beat-buesser  Yes, it is included in my original post.

**Reply by dwight-nwaigwe** [reply]
@beat-buesser  Here is another attempt at attaching the layer weights. I attached it as a .gz file. It's actually a .npz file, but I renamed the extension so that github allows it. Upon downloading, I suggest renaming the extension to .npz so that you can open it using Python tools as intended. [layer_weights.gz](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/8512840/layer_weights.gz)

**Reply by beat-buesser** [reply]
@animalcroc I cannot see the code where you calculate the accuracy of 97% using `model.predict`. Could you please show where?

**Reply by dwight-nwaigwe** [reply]
@beat-buesser I reposted the piece of code. In my original post, the formatting is strange, with big bold headlines. Maybe that's why you missed it. Here it is:

`#check that my model is accurate
for ind in range(0,10):
x_sample1=x_test[ind,:,:,:]
x_sample1=np.expand_dims(x_sample1,0)
mod_output=model.predict(x_sample1)
print(np.argmax(mod_output))
print(np.argmax(y_test[ind]))
`

**Reply by beat-buesser** [reply]
@animalcroc Ok, I see. Let me try out the weights you have shared. How did you get to the 97% accuracy reported earlier?

**Reply by dwight-nwaigwe** [reply]
@beat-buesser  Whoops, i responded outside this thread.

**Reply by beat-buesser** [reply]
@animalcroc No problem, I'll continue responding in this thread.

**Reply by beat-buesser** [reply]
Hi @animalcroc It looks like you also need to set the weights again after creating the graph and running the global initialiser in the 2 lines after `# New Addition: Update weights on graph`:

```python
import tensorflow.compat.v1 as tf
import numpy as np
tf.compat.v1.disable_eager_execution()
#from tensorflow.compat.v1.keras.models import Model
import tensorflow.compat.v1.keras.layers
from tensorflow.compat.v1.keras.models import Sequential
from tensorflow.compat.v1.keras.layers import Dense, Dropout, Activation, Flatten
from art.estimators.classification import TensorFlowClassifier
from art.utils import load_mnist
import pickle


# Step 1: Load the MNIST dataset

(x_train, y_train), (x_test, y_test), min_pixel_value, max_pixel_value = load_mnist()

# load layer weights
lay_weights = []
for i in range(12):
    lay_weights.append(np.load("/tmp/layer_weights/arr_" + str(i) + ".npy"))

    

# Step 2: Create the model
input_ph = tf.placeholder(tf.float32, shape=[None, 28, 28, 1])
labels_ph = tf.placeholder(tf.int32, shape=[None, 10])

model = Sequential()
model.add(Flatten(input_shape=(28, 28, 1)))
model.add(Activation('relu'))
model.add(Dense(300))
model.add(Activation('relu'))
model.add(Dense(200))
model.add(Activation('relu'))
model.add(Dense(150))
model.add(Activation('relu'))
model.add(Dense(150))
model.add(Activation('relu'))
model.add(Dense(100))
model.add(Activation('sigmoid'))
model.add(Dense(10))
model.layers[2].set_weights([lay_weights[0], lay_weights[1]])
model.layers[4].set_weights([lay_weights[2], lay_weights[3]])
model.layers[6].set_weights([lay_weights[4], lay_weights[5]])
model.layers[8].set_weights([lay_weights[6], lay_weights[7]])
model.layers[10].set_weights([lay_weights[8], lay_weights[9]])
model.layers[12].set_weights([lay_weights[10], lay_weights[11]])

#set logits equal to model output
logits=model(input_ph)

#check that my model is accurate
for ind in range(0,10):
    x_sample1=x_test[ind,:,:,:]
    x_sample1=np.expand_dims(x_sample1,0)
    mod_output=model.predict(x_sample1)
    print(np.argmax(mod_output), np.argmax(y_test[ind]))

sess = tf.Session()
sess.run(tf.global_variables_initializer())

# New Addition: Update weights on graph
update_ops = [tf.assign(to_var, from_var) for to_var, from_var in zip(model.trainable_weights, lay_weights)]
sess.run(update_ops)

# Step 3: Create the ART classifier

classifier = TensorFlowClassifier(
    clip_values=(min_pixel_value, max_pixel_value),
    input_ph=input_ph,
    output=logits,
    labels_ph=labels_ph,
    learning=None,
    sess=sess,
    preprocessing_defences=[],
)

# Step 5: Evaluate the ART classifier on benign test examples
predictions = classifier.predict(x_test)
accuracy = np.sum(np.argmax(predictions, axis=1) == np.argmax(y_test, axis=1)) / len(y_test)
print("Accuracy on benign test examples: {}%".format(accuracy * 100))
```

**Reply by dwight-nwaigwe** [reply]
@beat-buesser Great, thanks for the fix.

**Comment by dwight-nwaigwe** [comment]
[ART_error_pretrained.txt](https://github.com/Trusted-AI/adversarial-robustness-toolbox/files/8515703/ART_error_pretrained.txt)
I attached a text file with my code. Hopefully it's easier to read than the inline code I pasted in my initial post. The 97% accuracy is an estimate based on the number of times the model output matches the test output (you can also see this in my attached code). For your convenience, i repaste it here inline:

`#check that my model is accurate
for ind in range(0,10):
x_sample1=x_test[ind,:,:,:]
x_sample1=np.expand_dims(x_sample1,0)
mod_output=model.predict(x_sample1)
print(np.argmax(mod_output))
print(np.argmax(y_test[ind]))`
